rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -787,6 +795,29 @@ void PlayerImpl::delete_on_play_message_callback(const callback_handle_t & handl
     });
 }
 
+bool PlayerImpl::wait_for_sent_service_requests_to_finish(
+  const std::string & service_name,
+  std::chrono::duration<double> timeout)
+{
+  bool is_requests_complete = true;
+  if (!service_name.empty()) {
+    auto service_event_name = rosbag2_cpp::service_name_to_service_event_topic_name(service_name);
+    auto client_iter = service_clients_.find(service_event_name);
+    if (client_iter != service_clients_.end()) {
+      is_requests_complete = client_iter->second->wait_for_sent_requests_to_finish(timeout);
```

Comment:
Finally, it also will call PlayerServiceClientManager::wait_for_all_futures().   
So this operation does not wait for the future of the client of the specified service to complete, but waits for all futures to complete.

Reply:
@Barry-Xu-2018 In this particular case it will wait for all futures for specific client.
However if `service_name` is empty it will wait for all futures from all clients, - does it make sense?

Reply:
@MichaelOrlov 

> However if service_name is empty it will wait for all futures from all clients, - does it make sense?

Yes. It makes sense. 

> In this particular case it will wait for all futures for specific client.

Although the code retrieves the corresponding client based on the service name, `client->wait_for_sent_requests_to_finish()` actually calls `PlayerServiceClientManager::wait_for_all_futures()`. This function does not wait for all futures of a specific client; instead, it waits for all futures of all registered clients.